File: src/sorting_algo/MergeSort.java
package sorting_algo;

public class MergeSort {
    public static void main(String args[]){
        int arr[] = {90, 23, 101, 45, 65, 12, 11, 13, 5, 6, 7, 28, 67, 89, 34, 29};

        sorting(arr, 0, arr.length-1);
        System.out.print("Sorted array: ");
        SortedArrayPrint(arr);
    }

    static void merge(int[] arr,int low, int mid, int high){
        int n1 = mid - low +1;
        int n2 = high - mid;

        // temp array
        int arr1[] = new int[n1];
        int arr2[] = new int[n2];

        for (int i = 0; i < n1; i++)
            arr1[i] = arr[low+i];
        for (int j = 0; j < n2; j++)
            arr2[j] = arr[mid + 1 + j];

        int i=0, j=0, k=low;

        while (i<n1 && j<n2){
            if(arr1[i] <= arr2[j]){
                arr[k] = arr1[i];
                i++;
            }else {
                arr[k] = arr2[j];
                j++;
            }
            k++;
        }

        while (i<n1){
            arr[k] = arr1[i];
            k++;
            i++;
        }

        while (j<n2){
            arr[k] = arr2[j];
            j++;k++;
        }

    }

    static void sorting(int[] arr, int low, int high){
        if(low < high){
            int mid = (low+high)/2;

            sorting(arr, low, mid);
            sorting(arr,mid+1, high);
            merge(arr, low, mid, high);
        }
    }
    static void SortedArrayPrint(int[] arr){
        for (int i = 0; i < arr.length; i++) {
            System.out.print(arr[i] + " ");
        }
        System.out.println();
    }
}

File: MergeSort.js
const MergeSort = () => {
  let arr = [73, 57, 49, 99, 133, 20, 1, 64, 34, 25, 12, 22, 11, 90];
  sorting(arr, 0, arr.length - 1);
  console.log("Sorted Array:", arr);
};

const sorting = (arr, l, r) => {
  if (l < r) {
    let mid = Math.floor((l + r) / 2);
    sorting(arr, l, mid);
    sorting(arr, mid + 1, r);
    return merge(arr, l, mid, r);
  }
};

const merge = (arr, l, m, r) => {
  let n1 = m - l + 1;
  let n2 = r - m;

  let ar1 = new Array(n1);
  let ar2 = new Array(n2);

  for (let i = 0; i < n1; i++) ar1[i] = arr[l + i];
  for (let i = 0; i < n2; i++) ar2[i] = arr[m + 1 + i];

  let k = l,
    i = 0,
    j = 0;

  while (i < n1 && j < n2) {
    if (ar1[i] <= ar2[j]) {
      arr[k] = ar1[i];
      i++;
    } else {
      arr[k] = ar2[j];
      j++;
    }
    k++;
  }

  while (i < n1) {
    arr[k] = ar1[i];
    i++;
    k++;
  }

  while (j < n2) {
    arr[k] = ar2[j];
    j++;
    k++;
  }
  //   console.log([...arr]);
  //   return (newArr = [...arr]);
  //   console.log(arr);
};

MergeSort();
